astacus/coordinator/plugins/cassandra/plugin.py

```diff
@@ -95,22 +114,27 @@ def get_restore_steps(self, *, context: OperationContext, req: ipc.RestoreReques
             base.BackupNameStep(json_storage=context.json_storage, requested_name=req.name),
             base.BackupManifestStep(json_storage=context.json_storage),
             restore_steps.ParsePluginManifestStep(),
+            base.MapNodesStep(partial_restore_nodes=req.partial_restore_nodes),
         ] + cluster_restore_steps
 
     def get_restore_schema_from_snapshot_steps(self, *, context: OperationContext, req: ipc.RestoreRequest) -> List[Step]:
+        assert self.nodes is not None
+
         return [
             base.RestoreStep(storage_name=context.storage_name, partial_restore_nodes=req.partial_restore_nodes),
-            CassandraSubOpStep(op=ipc.CassandraSubOp.restore_snapshot_with_schema),
-            restore_steps.StartCassandraStep(partial_restore_nodes=req.partial_restore_nodes, override_tokens=True),
+            CassandraRestoreSubOpStep(op=ipc.CassandraSubOp.restore_snapshot_with_schema),
+            restore_steps.StopReplacedNodesStep(partial_restore_nodes=req.partial_restore_nodes, cassandra_nodes=self.nodes),
+            restore_steps.StartCassandraStep(replace_backup_nodes=True, override_tokens=True, cassandra_nodes=self.nodes),
             restore_steps.WaitCassandraUpStep(duration=self.restore_start_timeout),
         ]
 
     def get_restore_schema_from_manifest_steps(self, *, context: OperationContext, req: ipc.RestoreRequest) -> List[Step]:
+        assert self.nodes is not None
         client = CassandraClient(self.client)
 
         return [
             # Start cassandra with backed up token distribution + set schema + stop it
-            restore_steps.StartCassandraStep(partial_restore_nodes=req.partial_restore_nodes, override_tokens=True),
+            restore_steps.StartCassandraStep(override_tokens=True, cassandra_nodes=self.nodes),
```

Comment:
nit. This is not urgent, but there are some benefits to not reusing the same class in the steps list, even when not using their result :
the class name is used as a metric name so monitoring about steps duration and failures can be confusing if two steps have the same name (and communication between operators).

It can be helpful to have small subclasses of `StartCassandraStep` that just pin the booleans values and have names that describe which start it is.